README.rst
django-modelhandler
*******************

.. image:: https://badge.fury.io/py/django-modelhandler.svg
    :target: https://badge.fury.io/py/django-modelhandler

.. image:: https://travis-ci.org/sashgorokhov/django-modelhandler.svg?branch=master
    :target: https://travis-ci.org/sashgorokhov/django-modelhandler

A python logging handler that saves logs into django model. That's it.

Installation
============

Via pip:

.. code-block:: shell

    pip install django-modelhandler

Usage
=====

Add ``modelhandler`` to your INSTALLED_APPS, then, configure log handler in your desired way.
For example, using dictConfig:

.. code-block:: python

    {
        'version': 1,
        'handlers': {
            'modelhandler': {
                'class': 'modelhandler.handlers.LogModel',
                'level': 'ERROR'
            }
        },
        'loggers': {
            'my_logger': {
                'handlers': ['modelhandler'],
                'level': 'ERROR'
            }
        }
    }

Run migrations that will create a ``Log`` model:

.. code-block:: shell

    python manage.py migrate

And now you can start logging in django model.

Getting logs:

.. code-block:: python

    from modelhandler.models import Log
    # Get the latest log
    log = Log.objects.latest()
    log.name # logger name
    log.level # logging level integer
    log.levelname # logging level as string (DEBUG, INFO, etc.)
    log.message # the log message
    log.traceback # traceback, if exists. default: None
    log.filename # filename (with ext) where the log was sent
    log.funcName # function name where the log was sent
    log.created # log creation datetime
    log.formatted # the log message as if it was written in file. (with [datetime] [level] etc.)


If you have a django admin enabled, then you can browse your logs on model Log of application modelhandler.
It has a customized modeladmin to enabale filtering by logger name and levelname, and searching by message.

If you would like to customize a log model (to alter models Meta), then just subclass a ``modelhandler.models.Log``
model, do whatever you want and add your model to LogModel handler parameters:

.. code-block:: python

    'handlers': {
        'modelhandler': {
            'class': 'modelhandler.handlers.LogModel',
            'model': 'path.to.your.model'
            'level': 'ERROR'
        }
    }

If you using celery in your project then you might want to add some model cleaning tasks in ``CELERYBEAT_SCHEDULE``:

.. code-block:: python

    CELERYBEAT_SCHEDULE = {
        'cleanup_day': {
            'task': 'modelhandler.tasks.cleanup_day',
            'schedule': timedelta(days=1)
        }, # OR
        'cleanup_week': {
            'task': 'modelhandler.tasks.cleanup_week',
            'schedule': timedelta(days=7)
        }, # OR
        'cleanup_month': {
            'task': 'modelhandler.tasks.cleanup_month',
            'schedule': timedelta(days=30)
        }
    }

``modelhandler.tasks.cleanup_day`` will delete all logs that are older than one day from time of task execution.
``modelhandler.tasks.cleanup_week`` and ``modelhandler.tasks.cleanup_month`` are similar.

If you want to customize the time of deletion, there is a task ``modelhandler.tasks.cleanup_logs`` that accepts
a ``before`` parameter that must be a datetime object or None (in this case a value of timezone.now() will be taken).
There is no magic: just ``Log.objects.filter(created__lte=before).delete()``


HISTORY.rst
.. :changelog:

History
-------

1.0.0 (2016-06-12)
++++++++++++++++++

* Initial release

tests/tests.py
import logging
import os

from django.test import TestCase
from modelhandler.handlers import LogModel
from modelhandler.models import Log
from modelhandler.utils import LEVEL_TO_NAME


class TestModelHandler(TestCase):
    def setUp(self):
        self.handler = LogModel(model=Log, level=logging.DEBUG)
        self.logger = self.get_logger(self.__class__.__name__)

    def get_logger(self, name="", level=logging.DEBUG, handler=None):
        logger = logging.getLogger(name)
        logger.setLevel(level)
        logger.handlers = list()
        logger.addHandler(handler or self.handler)
        return logger

    def test_name(self):
        self.logger.info("")
        self.assertTrue(Log.objects.filter(name=self.logger.name).exists())

    def test_root_logger_name(self):
        logger = self.get_logger()
        logger.info("")
        self.assertTrue(Log.objects.filter(name=logger.name).exists())

    def test_levels(self):
        for level, levelname in LEVEL_TO_NAME.items():
            if not self.logger.isEnabledFor(level):
                continue
            self.logger.log(level, levelname)
            log = Log.objects.latest()
            self.assertEqual(log.level, level)
            self.assertEqual(log.levelname, levelname)

    def test_traceback(self):
        message = 'My Message'
        try:
            raise ValueError(message)
        except:
            self.logger.exception("")
        log = Log.objects.latest()
        self.assertTrue(log.traceback)
        self.assertIn(message, log.traceback)

    def test_filename(self):
        self.logger.debug("")
        filename = os.path.split(__file__)[-1]
        if filename.endswith('pyc'):
            filename = filename[:-1]
        self.assertTrue(Log.objects.filter(filename=filename).exists())

    def test_funcname(self):
        self.logger.debug("")
        self.assertTrue(Log.objects.filter(funcName='test_funcname').exists())

    def test_message(self):
        message = 'Hello, %s'
        param = 'world'
        self.logger.debug(message, param)
        log = Log.objects.latest()
        self.assertEquals(log.message, message % param)

tests/settings.py
import os
import sys

BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
sys.path.append(os.path.dirname(BASE_DIR))
SECRET_KEY = '2a@*ye6$(^h5tqsdz(ga!=_gznq#xpox3c7s&h4hq!3#rbds3@'
DEBUG = True

INSTALLED_APPS = [
    'modelhandler',
]

DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': os.path.join(BASE_DIR, 'db.sqlite3'),
    }
}

LOGGING = {
    'version': 1,
    'handlers': {
        'log_model': {
            'class': 'modelhandler.handlers.LogModel',
            'level': 'DEBUG'
        }
    },
    'loggers': {
        'root': {
            'handlers': ['log_model'],
            'level': 'DEBUG'
        },
    }
}

modelhandler/models.py
import logging
from django.db import models
from modelhandler.utils import LEVEL_TO_NAME


class Log(models.Model):
    name = models.CharField(max_length=255)
    level = models.PositiveIntegerField(default=0, choices=LEVEL_TO_NAME.items())
    message = models.TextField()
    traceback = models.TextField(default=None, null=True, blank=True)
    filename = models.CharField(max_length=255, default=None, null=True, blank=True)
    funcName = models.CharField(max_length=255, default=None, null=True, blank=True)
    created = models.DateTimeField()
    formatted = models.TextField()

    class Meta:
        ordering = ['created']
        get_latest_by = 'created'

    @property
    def levelname(self):
        return logging.getLevelName(self.level)

    def __str__(self):
        return self.message


modelhandler/handlers.py
import datetime
import logging

from django.utils.module_loading import import_string

__all__ = ['LogModel']


class LogModel(logging.Handler):
    def __init__(self, model=None, *args, **kwargs):
        self.model = model
        super(LogModel, self).__init__(*args, **kwargs)

    def get_model(self):
        if self.model and isinstance(self.model, str):
            self.model = import_string(self.model)
        if not self.model:
            from modelhandler.models import Log
            self.model = Log
        return self.model

    def emit(self, record):
        """
        :param logging.LogRecord record:
        """
        instance = self.get_model()()
        instance.name = record.name[:255]
        instance.level = record.levelno
        instance.message = record.getMessage()
        instance.created = datetime.datetime.fromtimestamp(record.created)
        instance.filename = record.filename[:255] if isinstance(record.filename, str) else None
        instance.funcName = record.funcName[:255] if isinstance(record.funcName, str) else None
        instance.formatted = self.format(record)
        if record.exc_info:
            formatter = self.formatter or logging.Formatter()
            instance.traceback = formatter.formatException(record.exc_info)
        instance.save()


modelhandler/migrations/0001_initial.py
# -*- coding: utf-8 -*-
# Generated by Django 1.9.7 on 2016-06-11 19:39
from __future__ import unicode_literals

from django.db import migrations, models


class Migration(migrations.Migration):
    initial = True

    dependencies = [
    ]

    operations = [
        migrations.CreateModel(
            name='Log',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('name', models.CharField(max_length=255)),
                ('level', models.PositiveIntegerField(default=0)),
                ('message', models.TextField()),
                ('traceback', models.TextField(blank=True, default=None, null=True)),
                ('filename', models.CharField(blank=True, default=None, max_length=255, null=True)),
                ('funcName', models.CharField(blank=True, default=None, max_length=255, null=True)),
                ('created', models.DateTimeField()),
                ('formatted', models.TextField()),
            ],
            options={
                'get_latest_by': 'created',
                'ordering': ['created'],
            },
        ),
    ]
